### ISSUE TITLE
Findings after the summer season

### ISSUE THREAD

**Issue by Timo-Fi** [forrestguice/SuntimesWidget#625]
Hi, I've been using the version 0.14.7 throughout the summer, here are some aspects I'd like to mention.

1)  On my cheapo devices the font size of the Sun Position 3x1 widget is so small that the information ("alt", "noon alt", etc) are difficult to discern when outdoors under bright daylight. The Sun Position -tool shows the same information at same/similar font size but mostly with bold font style, these I can read more easier, but still with some difficulty.  Except for the 2nd row, the time stamp row, that is not in bold font face. So, I'm hoping if they could have somewhat larger font size, or perhaps there could even be a new widget or a new layout, so that the current altitude angle of the sun could be shown at very large font size, like taking the whole area of a 2x1 widget or even at 4x2 size.

2) In the The Sun Position -tool, could it be possible, that a tap on the "Alt" (the current altitude angle of the sun)  would open an input dialog where one could enter a desired altitude angle, and the dialog would have two buttons named as "Rising" and "Setting" or such, either one would close the input dialog and then the output of the tool would be set according to that altitude? This would make it more easier to inspect a particular situation than using the animation function.  

3) A very minor plea... In the Sun Position 3x2 widget, could there be a new layout to show the altitude angle of the sun as a curve line on a chart?  X-axis as the time from -12h to +12h and Y-axis as the altitude angle from 0 to 90 degrees, with hairline divisions for both scales, and the path of the sun as a yellow curve line with the icon of sun at the current time. This would show a great overall info at a glance, and it would be a nice conversation feature, as by changing the Location -setting one could quickly compare the sunbathing conditions at different destinations. 

4) I have set one of the widgets to the perform: "On Tap: Update Widget" -action. I'm thinking if this could be changed to update all the existing widgets simultaneously, as the non-updated state never(?) is desirable.

5) In the Moon Widget (3x1) Major Phases (that I show in 4x1 size) the long month names (like September) push the time of the clock down to a new line,  so only the top half of that line is visible  in 4x1. When I allocate 4x2 then that line is visible but a lot of real estate is wasted.  Could the months be shown with their thee-first-letters abbreviation?     

6) Most likely I'm the only one with this, but anyway.... Could it be at all possible to have it so that the notifications that are iss7ued by the Suntimes would be acted upon by Android OS just like it happens now, but after that they would simply sink without a trace? This would allow notifications to run indefinitely. Currently the OS will stop notifications from an app when there are about 20 of them collected in the notification_shade and then the user has to remove them. So, I'd like to have a device running unattended on the wall, showing a nice screen with Sun Position Widget, Suntimes Widget, Moon Widget while it also speaks out the hourly notifications.  
 
Thank you so very much for the very best of the best apps,
Timo

**Comment by forrestguice** [comment]
Thanks @Timo-Fi. This is great feedback. ❤️
I think most can be accommodated. I'll review these again and comment on the specifics later.

I think number 4 is reasonable behavior (and an easy change). Number 6 sounds like what another user calls [Ephemeral notifications)[https://github.com/forrestguice/SuntimesWidget/issues/552], which is a definite for `v0.15.0` (code has been written). I'm unsure how the app should label that feature though. While those notifications are "ephemeral" or "transient", a shorter label is preferable. Its currently labeled "quick notification", an addition to the "alarm" and "notification" types.

**Comment by Timo-Fi** [comment]
Hello **@forrestguice**, great to hear that it is already in consideration (I did not now the about  the word 'ephemeral' at all, so I never read the discussion).  Sure, it can add confusion for the user since it is a brand new type.  For a shorter label, something like 'poke'  or 'note' would direct me to seek more information.  On the other hand, the second dialog of the task of alarm/notification creation seems to have room for one extra row, with a checkbox and a title and like 'non-accumulative', this way it likely could apply for both alarms and notifications.

**Comment by Timo-Fi** [comment]
Hello again,

Last night the damned DST ended here, on this year it was in effect from 27th. Feb. to 30th. Oct.

During yesterday, on one OS7 device and on one OS12 device (that are usually not connected to internet) all the Notifications were triggered normally. Today none of the 23 'Every Day' Notifications are triggered.  In the Alarms list the status for each of them are either '~6d Hh from now' or '~5d Hh from now' (where H is a value from 0 to 23). They are still in this condition, I have not connected them to internet but even then the clock time has changed correctly, showing now the DST=0 time.  I'll upload screenshots if some kind of views are needed. 

At the same time, on one OS11 device, that normally is connected to the internet, the above did not happen.  However, some (likely three) days ago I reloaded the saved Notifications -list to this device because one of the Notifications was not triggered,  that Notification had the '~Dd Hh from now' status (instead of '~Hh Mm from now'). I did not check the status of the other Notifications on this device. On this device the gmail and calendar are synced to Google.

In addition, on one OS6 device triggering of the Notifications stopped 165 days ago. In anticipation of self-recovery, nothing has been done towards recovering functionality (but the device has since been rebooted at leas twice due to other reasons). This device is not normally connected to the internet.  Today it is still not triggering, the status of the Notifications are  '~165d Hh from now' or '~164d Hh from now'. 

The devices that are not normally connected to the internet are briefly connected, something like once in month or two, in order to sync the time. Nothing from these devices are backed/synced to any network service.

All the devices are rebooted once in two months as a routine, and additionally in case when some app does not behave. I have never encountered a problem from the Suntimes app that would need a rebooting.  

Perhaps Google has its spying nose somewhere there in the Alarim/Notifications routine and that is messing things up.  I'm thinking that the Suntimes could possibly detect/correct such situations as the status of each Notification could be compared against their definitions in the Alarm list.

**Comment by forrestguice** [comment]
> DST ended here, on this year it was in effect from 27th. Feb. to 30th. Oct.

It's troubling those notifications failed to trigger. It can't be coincidence that they worked reliably for some time and then failed after DST ended.

I think the relationship between the bug and internet connectivity is probably in the system's "automatic date and time" and "automatic timezone". Those settings are backed by network (ntp server). I tried disabling them and then manually setting the date/time to a minute before the DST boundary. I waited but the timezone didn't switch over on its own (remained in DST), only when the time was manually set. I'm reading from various sources that those "automatic" settings are required to have DST change automatically.

I figure there is a DST related bug somewhere in the scheduling code (but maybe its somewhere else).. I'm having difficulty finding it (will continue looking), but this scenario isn't well tested. There isn't DST where I live, so I might fail to notice these bugs.

> I'm thinking that the Suntimes could possibly detect/correct such situations as the status of each Notification could be compared against their definitions in the Alarm list.

Its definitely detectable, but I think it requires looping over and examining each alarm in the database. For that reason I'm not sure where/when it should be run. I'm hesitant to run it just anywhere supposing it takes some time (and introduces a delay). For example, rebooting should reschedule all alarms (but since it didn't recover your OS6 device, that's something I'm going to look closely at).

**Comment by Friptick** [comment]
>https://github.com/forrestguice/SuntimesWidget/issues/552, which is a definite for v0.15.0 ([code](https://github.com/forrestguice/SuntimesWidget/tree/alarm-type-qnotification) has been written).

YESSSS!! So looking forward to it, thanks. The notifications pile-up is driving me crazy (literally came back here to bump that issue, _as politely and gratefully as possible, of course_. :)

>I'm unsure how the app should label that feature though. While those notifications are "ephemeral" or "transient", a shorter label is preferable. Its currently labeled "quick notification", an addition to the "alarm" and "notification" types.

"Quick notification" is understandable. I believe "ephemeral" (fleeting, impermanent, temporary) is the most precise term but @Timo-Fi didn't get it so your idea is better.

>It's troubling those notifications failed to trigger.

Yes I also have had [a similar issue](#602) (the "next alarm" was literally showing a negative time value) but the good news here is that for whatever reason it went away. Can confirm that the recent daylight-savings transition went very smoothly, on a device which had no network connectivity at all.

**Comment by Timo-Fi** [comment]
>rebooting should reschedule all alarms (but since it didn't recover your OS6 device

When I reboot in the attempt of recovering a stuck Notification the reboot never recovers, this is so with all my devices. Importing the stored Alarms.txt always recovers such situation. In the past I used to recover them by toggling the state of such Notifications OFF/ON, it always helped. I do not do this anymore because every Notification has to be individually checked and then toggled if stuck, loading the Alarms.txt is easier/faster.

> 'm hesitant to run it just anywhere supposing it takes some time (and introduces a delay). 

Surely understandable, particularly if the operation of Importing the Alarms.txt at all indicates the processing time required.  With my 23 Notifications, after Clearing them all and then Importing, it takes five or six heartbeats (nearly an indefinite amount of time) with the screen pitch black until the Notifications appears.  

The  status of the Notifications (on the OS7 device) are now  '~9d HHh from now' or '~10d HHh from now'.  The amount of d(ays) have been increasing one per day. So like [Friptick](https://github.com/Friptick) mentioned, the issue could possibly be a result of a negative value that in some phase gets taken as absolute value. Perhaps the negative values somehow were brewed in this occasion because this DST change took the time (of the clock) "backwards".  It might also be a hint of some sort that the status of the Notifications jumped to ~6d / ~5d (from now), as if a weekly routine could be in play (all my 23 Notifications are however set as "Every Day"). 

About the "ephemeral" (as the name/tittle for the vanishing Alarms/Notifications)  as mentioned  [Friptick](https://github.com/Friptick) I just like to say that it is perfectly good choice too. I believe everyone will look it up from a dictionary when a strange word appears in the GUI of a software thar is being used, it is a different case when such word is encountered on a discussion forum/newspaper where it is easy to just skip to the next one.  

BTW: With the Suntimes app my devices have massive number of Notification instanced, 8395 pcs/year from Suntimes app alone.  Before the Suntimes app I only had very small number of Alarm instances, something like less than 90 pcs/year of stray Alarms (that I still create as the need may be) and 365 pcs/year of morning Alarms.  And notably, before the Suntimes app I encountered some situations where an Alarm was not triggered,I was not able to associate these cases with any particular cause, but I got the quite a strong inkling that they might have occurred due to not rebooting the device frequently enough.

**Comment by forrestguice** [comment]
> When I reboot in the attempt of recovering a stuck Notification the reboot never recovers, this is so with all my devices. 

I went looking for a problem with the reboot code and found one. I discovered that alarms/notifications that were already in the "sounding" or "snoozing" state were skipped over (due to a bug in the transition code), so the reboot failed to reschedule these. It is actually a little worse for notifications than alarms, because they don't time out and aren't rescheduled right away. The notifications that pile up in the tray remain in the "sounding" state until clicked (or removed). If the device is powered off with notifications in the tray, those notifications are still "sounding" when the device reboots. This is likely just one way to create stale/zombie notifications - it was only after noticing the bug in the transition code that I realized it might be triggered this way.

This should be fixed with `v0.14.10` and a reboot really will fix those stale alarms/notifications.

> the issue could possibly be a result of a negative value that in some phase gets taken as absolute value

I think what we see afterward (negative values) is a separate issue with the display code. It is showing an absolute value, something like `abs(alarmtime - now)` (but I can't remember why exactly). The `alarmtime` is no longer updated though (so now counting up over time).

> somehow were brewed in this occasion because this DST change took the time (of the clock) "backwards".

I searched the code for potential daylight saving issues but couldn't find anything that would affect the repeating "event" alarms. These are all scheduled by UTC and so should be immune. The regular "clock time" alarms are affected by DST though (and somebody recently reported their alarm went off an hour late). The latest patch (`v0.14.10`) listens for time zone changes (which should reschedule after dst changes too). 

So I'm still not sure why those notifications failed earlier. I have discovered ways to /make/ them fail though (just ignore notifications for a day, followed by a reboot). Maybe that's progress. 

btw, the latest patch also includes a few changes mentioned here. Its now possible to "update all widgets", and configure widgets to show abbreviated month names. The 3x1 sun position labels are now bold, and checking "scale text and icons" now scales them. That should help some with readability issues.

**Comment by forrestguice** [comment]
> The notifications pile-up is driving me crazy (literally came back here to bump that issue, as politely and gratefully as possible, of course. :)

@Friptick I swear time passes quicker with age (and then suddenly you realize a lot of it has gone by). I realize its been quite awhile since I promised that feature - sorry about that. That code has just been "hanging out" in a branch while I work on other "big" stuff (which then "hangs out" too). It has recently gotten very cold, so I'm suddenly finding it a lot easier to spend time in front of a screen. That's good news for the release cycle ;)

**Comment by Friptick** [comment]
That's great news @forrestguice ! No pressure, we appreciate everything you do. :smile:

**Comment by Timo-Fi** [comment]
Great! This happened:  I updated my devices to Suntimes version 0.14.10. On the Android 7 device that I have left the the Notifications at stalled state for a while, I inspected the Notifications list right after the update, and they were still stalled (that is to say I saw the  '~Dd Hh from now' -status instead of the normal '~Hh Mm from now' -status). Then I rebooted the device and it was still the same. Then I left the device on it's own with power On (it's just sitting on a table 24/7 showing widgets). Then, to my immense surprise, on next day, it was very likely on the 25th hour since the update, the device briskly spoke the hourly Notification correctly, and has been doing so since then. So now all the 23 Notifications plays OK.

**Comment by forrestguice** [comment]
That delay is pretty odd. I know that after reboot it sometimes takes several minutes before Suntimes gets to run though - I'm guessing the list must have been opened before Suntimes got its opportunity.

> I inspected the Notifications list right after the update

I don't actually expect a change immediately after each update. In fact the `PACKAGE_REPLACED` event might be something that breaks alarms on some devices. That isn't my personal experience, but I wonder if the app also needs to reschedule alarms after updating too.

**Comment by forrestguice** [comment]
> Then I rebooted the device and it was still the same. 

I think this touches on another issue, which is the reboot code doesn't indicate when its running (or has completed). I think it should probably show a notification while this is in progress. When I reboot I see a lot of "boot notifications" that are shown for only a moment and then disappear, so this seems like expected behavior.
